File: packages/sdk-ruby/lib/melaya/http_client.rb
# frozen_string_literal: true

require "net/http"
require "uri"
require "json"
require "openssl"

require_relative "errors"

module Melaya
  # Internal HTTP client. Supports Bearer JWT *and* mk_* platform API key.
  # The credential is sent ONLY via the Authorization header — never in the URL
  # query string, so it cannot leak into access logs or proxies. TLS is
  # enforced by default; never log secrets.
  #
  # Retry policy: bounded exponential back-off with jitter on network errors,
  # 429, and 5xx — but ONLY for idempotent GET requests (max 2 retries).
  # POST/PUT/PATCH/DELETE are never retried. Retry-After header is honoured
  # on 429. Per-request timeout default: 30 seconds (configurable).
  class HttpClient
    DEFAULT_BASE_URL    = "https://api.melaya.org"
    DEFAULT_TIMEOUT_MS  = 30_000  # milliseconds

    # Maximum additional retries after the first attempt (2 retries = 3 total
    # attempts) for idempotent GET requests only.
    MAX_GET_RETRIES = 2
    RETRY_STATUSES  = [429, 500, 502, 503, 504].freeze

    # @param api_key [String] mk_* platform key or Bearer JWT
    # @param base_url [String]
    # @param verify_ssl [Boolean]
    # @param timeout_ms [Integer] per-request timeout in milliseconds (default 30 000)
    def initialize(api_key:, base_url: DEFAULT_BASE_URL, verify_ssl: true,
                   timeout_ms: DEFAULT_TIMEOUT_MS)
      raise ArgumentError, "Melaya: TLS certificate verification cannot be disabled." unless verify_ssl
      # Never store in a way that could leak to logs accidentally — keep as
      # an opaque token string only accessible through the private accessor.
      @_tok       = api_key.freeze
      @base_uri   = URI.parse(base_url.chomp("/"))
      @verify_ssl = true
      @timeout_s  = (timeout_ms / 1000.0).ceil
    end

    # ── Public verb helpers ────────────────────────────────────────────────────

    def get(path, params = {})
      request(:get, path, params: params)
    end

    def post(path, body = nil)
      request(:post, path, body: body)
    end

    def put(path, body = nil)
      request(:put, path, body: body)
    end

    def patch(path, body = nil)
      request(:patch, path, body: body)
    end

    def delete(path, params = {})
      request(:delete, path, params: params)
    end

    private

    def build_uri(path, params = {})
      uri = URI.parse("#{@base_uri}#{path}")
      # SECURITY: the credential is never placed in the query string; it is
      # sent only via the Authorization header (see make_request).
      query = {}
      params.each { |k, v| query[k.to_s] = v.to_s unless v.nil? }
      uri.query = URI.encode_www_form(query) unless query.empty?
      uri
    end

    def make_request(method, uri, body)
      req = case method
            when :get    then Net::HTTP::Get.new(uri)
            when :post   then Net::HTTP::Post.new(uri)
            when :put    then Net::HTTP::Put.new(uri)
            when :patch  then Net::HTTP::Patch.new(uri)
            when :delete then Net::HTTP::Delete.new(uri)
            else raise ArgumentError, "Unknown HTTP method: #{method}"
            end

      # Authorization: never expose token in error output below
      req["Authorization"] = "Bearer #{@_tok}"
      req["Accept"]        = "application/json"
      req["User-Agent"]    = "melaya-ruby/#{Melaya::VERSION}"

      if body
        req["Content-Type"] = "application/json"
        req.body = JSON.generate(body)
      end

      req
    end

    def request(method, path, params: {}, body: nil)
      uri = build_uri(path, params)

      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl      = uri.scheme == "https"
      http.verify_mode  = OpenSSL::SSL::VERIFY_PEER
      http.open_timeout = @timeout_s
      http.read_timeout = @timeout_s

      # Only GET requests are retried (idempotent); all mutating verbs fail fast.
      retryable = (method == :get)
      attempt   = 0

      retry_after_hdr = nil
      begin
        attempt += 1
        retry_after_hdr = nil  # reset on each attempt
        req  = make_request(method, uri, body)
        resp = http.request(req)
        # Snapshot Retry-After before parse() consumes the response object,
        # so we can honour the header even after the MelayaError is raised.
        retry_after_hdr = resp["retry-after"] || resp["Retry-After"]
        parse(resp)
      rescue MelayaError => e
        if retryable && RETRY_STATUSES.include?(e.status) && attempt <= MAX_GET_RETRIES
          # Build a minimal resp-like object carrying only the header we need,
          # so _backoff_delay can honour Retry-After without holding the socket.
          hdr_carrier = { "retry-after" => retry_after_hdr }
          delay = _backoff_delay(attempt, e, hdr_carrier)
          sleep(delay)
          retry
        end
        raise
      rescue Errno::ECONNREFUSED, Net::OpenTimeout, Net::ReadTimeout
        raise unless retryable && attempt <= MAX_GET_RETRIES
        sleep(_backoff_delay(attempt, nil))
        retry
      end
    end

    # Exponential backoff with ±25 % jitter; honours Retry-After on 429.
    # Base: 2^(attempt-1) seconds, capped at 16 s before jitter.
    #
    # Retry-After resolution order (first match wins):
    #   1. HTTP `Retry-After` response header — seconds integer or HTTP-date
    #   2. JSON body `retryAfter` / `retry_after` field (legacy fallback)
    #   3. Exponential back-off
    def _backoff_delay(attempt, err, resp = nil)
      if err.is_a?(MelayaError) && err.status == 429
        # 1. HTTP Retry-After header (preferred, RFC 7231)
        if resp.respond_to?(:[]) && (ra_hdr = resp["retry-after"] || resp["Retry-After"])
          secs = _parse_retry_after_header(ra_hdr)
          return [secs, 0.5].max if secs
        end

        # 2. JSON body fallback ("retryAfter" or "retry_after")
        if err.respond_to?(:body) && err.body.is_a?(Hash)
          ra = err.body["retryAfter"] || err.body["retry_after"]
          return [ra.to_f, 0.5].max if ra
        end
      end
      base  = [2**(attempt - 1), 16].min.to_f
      jitter = base * 0.25 * (rand - 0.5) * 2  # ±25 %
      [base + jitter, 0.1].max
    end

    # Parse an RFC 7231 Retry-After value: either a delay-seconds integer
    # or an HTTP-date string. Returns seconds as Float, or nil if unparseable.
    def _parse_retry_after_header(value)
      str = value.to_s.strip
      # Delay-seconds: plain non-negative integer
      if str =~ /\A\d+\z/
        return str.to_f
      end
      # HTTP-date (e.g. "Wed, 21 Oct 2099 07:28:00 GMT")
      begin
        require "time"
        target = Time.httpdate(str)
        delay  = target - Time.now
        return [delay, 0.0].max
      rescue ArgumentError, TypeError
        nil
      end
    end

    def parse(resp)
      text = resp.body.to_s.strip
      data = begin
        text.empty? ? nil : JSON.parse(text)
      rescue JSON::ParserError
        text
      end

      status = resp.code.to_i
      if status >= 400
        # Two error envelope shapes:
        #   1. { error: 'tier_insufficient', tier: '...' }  -> 403
        #   2. { error: '...', message: '...', code: '...' }
        # Extract error code safely — never echo raw body in message
        err_code = data.is_a?(Hash) ? data["error"] : nil

        if status == 403 && err_code == "tier_insufficient"
          raise TierInsufficientError.new(tier: data.is_a?(Hash) ? data["tier"] : nil, body: data)
        end
        if status == 429
          # Raised here; the GET retry loop above may swallow-and-retry it —
          # callers only see it once retries are exhausted.
          ra = _parse_retry_after_header(resp["retry-after"] || resp["Retry-After"])
          raise RateLimitError.new(retry_after: ra, body: data)
        end

        msg = "Melaya API #{resp.code}" + (err_code ? " (#{err_code})" : "")
        raise MelayaError.new(msg, status: status, code: err_code, body: data)
      end

      # The API may wrap payload in { "ok": false, ... } for request-level failures.
      if data.is_a?(Hash) && data["ok"] == false
        err_code = data["error"]
        msg = "Melaya API request failed" + (err_code ? ": #{err_code}" : "")
        raise MelayaError.new(msg, status: resp.code.to_i, code: err_code, body: data)
      end

      data
    end
  end
end

File: packages/sdk-php/src/HttpClient.php
<?php

declare(strict_types=1);

namespace Melaya;

/**
 * Internal cURL-based HTTP client.
 *
 * The credential (mk_* API key or session JWT) is sent ONLY via the
 * Authorization: Bearer header — never in the URL query string, so it cannot
 * leak into access logs, proxies, or referrer headers.
 *
 * Error mapping:
 *   - { "error": "tier_insufficient", "tier": "..." }  → 403  → MelayaException (code "tier_insufficient")
 *   - { "error": "...", "message": "...", "code": "..." } → 4xx/5xx → MelayaException
 *   - { "ok": false, ... } → MelayaException (envelope-level failure)
 *
 * Retry: bounded exponential backoff with jitter on GET requests only
 *   (attempt count configurable via the client's `retries` option, default 3)
 *   for network errors, 429, and 5xx. Retry-After header is honoured on 429.
 *   POST/PUT/PATCH/DELETE are never retried (not idempotent).
 * Timeout: per-request timeout (default 30 s, configurable). Passed as CURLOPT_TIMEOUT.
 * TLS: certificate and hostname verification are always enabled.
 * Security: the apiKey/JWT is NEVER logged or included in exception messages.
 */
class HttpClient
{
    /** Maximum backoff cap in seconds. */
    private const BACKOFF_CAP_SEC = 30;

    private readonly int  $maxRetries;
    private readonly int  $timeoutSec;

    public function __construct(
        /** @var string Bearer token — either mk_* API key or JWT; never logged. */
        private readonly string $bearer,
        private readonly string $baseUrl,
        /**
         * @deprecated No longer used. The API key is sent only via the
         * Authorization header; it is never injected into the query string.
         * Retained for backward-compatible construction.
         */
        private readonly bool   $injectApiKeyParam = true,
        int                     $timeoutSec        = 30,
        int                     $maxRetries        = 3,
    ) {
        $this->maxRetries = $maxRetries;
        $this->timeoutSec = $timeoutSec;
    }

    /** @return mixed */
    public function get(string $path, array $query = []): mixed
    {
        return $this->request('GET', $path, $query, null);
    }

    /** @return mixed */
    public function post(string $path, mixed $body = null): mixed
    {
        return $this->request('POST', $path, [], $body);
    }

    /** @return mixed */
    public function put(string $path, mixed $body = null): mixed
    {
        return $this->request('PUT', $path, [], $body);
    }

    /** @return mixed */
    public function patch(string $path, mixed $body = null): mixed
    {
        return $this->request('PATCH', $path, [], $body);
    }

    /** @return mixed */
    public function delete(string $path, array $query = []): mixed
    {
        return $this->request('DELETE', $path, $query, null);
    }

    // ── Internal ─────────────────────────────────────────────────────────────

    private function buildUrl(string $path, array $query): string
    {
        $base = rtrim($this->baseUrl, '/');
        $p    = '/' . ltrim($path, '/');
        // SECURITY: the credential is sent ONLY via the Authorization: Bearer
        // header (see request()). It must NEVER be added to the query string,
        // where it would leak into access logs, proxies, and referrers.
        $filtered = array_filter($query, static fn($v) => $v !== null && $v !== '');
        return $base . $p . (count($filtered) > 0 ? '?' . http_build_query($filtered) : '');
    }

    /** @return mixed */
    private function request(string $method, string $path, array $query, mixed $body): mixed
    {
        $url     = $this->buildUrl($path, $query);
        $headers = [
            'Authorization: Bearer ' . $this->bearer,
            'Accept: application/json',
            'User-Agent: melaya-php-sdk/0.2.0',
        ];

        $jsonBody = null;
        if ($body !== null) {
            $jsonBody  = json_encode($body, JSON_THROW_ON_ERROR | JSON_UNESCAPED_UNICODE);
            $headers[] = 'Content-Type: application/json';
            $headers[] = 'Content-Length: ' . strlen($jsonBody);
        }

        // Only GET requests are retried (idempotent). POST/PUT/PATCH/DELETE are
        // executed exactly once — retrying non-idempotent requests risks duplicates.
        $isIdempotent = ($method === 'GET');
        $maxAttempts  = $isIdempotent ? (max(0, $this->maxRetries) + 1) : 1;

        $attempt  = 0;
        $delaySec = 1.0;

        while (true) {
            $attempt++;
            [$status, $raw, $curlErr, $retryAfter] = $this->curlExec($method, $url, $headers, $jsonBody);

            $isNetworkErr  = ($curlErr !== '');
            $isRetryStatus = ($status === 429 || $status >= 500);

            if ($isIdempotent && ($isNetworkErr || $isRetryStatus) && $attempt < $maxAttempts) {
                // On 429 honour Retry-After if present; otherwise use exponential backoff + jitter.
                if ($status === 429 && $retryAfter > 0) {
                    $waitSec = min($retryAfter, self::BACKOFF_CAP_SEC);
                } else {
                    // jitter: uniform random in [delaySec/2 .. delaySec]
                    $waitSec = $delaySec * 0.5 + lcg_value() * $delaySec * 0.5;
                    $waitSec = min($waitSec, self::BACKOFF_CAP_SEC);
                }
                // sleep() takes integer seconds; usleep() for sub-second precision
                $waitUs = (int) round($waitSec * 1_000_000);
                usleep($waitUs);
                $delaySec = min($delaySec * 2, self::BACKOFF_CAP_SEC);
                continue;
            }

            if ($isNetworkErr) {
                throw new MelayaException('Melaya: network error (cURL)', 0);
            }

            return $this->parse($status, $raw);
        }
    }

    /**
     * Execute one cURL request.
     * @return array{int, string, string, int} [statusCode, rawBody, curlError, retryAfterSec]
     */
    private function curlExec(string $method, string $url, array $headers, ?string $jsonBody): array
    {
        $ch = curl_init($url);

        // CURLOPT_TIMEOUT is per-request wall-clock timeout in seconds.
        curl_setopt_array($ch, [
            CURLOPT_CUSTOMREQUEST  => $method,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT        => $this->timeoutSec,
            CURLOPT_SSL_VERIFYPEER => true,
            CURLOPT_SSL_VERIFYHOST => 2,
            CURLOPT_HTTPHEADER     => $headers,
            // Capture response headers so we can read Retry-After on 429
            CURLOPT_HEADERFUNCTION => static function ($ch, string $header) use (&$retryAfterRaw): int {
                if (str_starts_with(strtolower($header), 'retry-after:')) {
                    $retryAfterRaw = trim(substr($header, strlen('retry-after:')));
                }
                return strlen($header);
            },
        ]);

        if ($jsonBody !== null) {
            curl_setopt($ch, CURLOPT_POSTFIELDS, $jsonBody);
        }

        $retryAfterRaw = null;
        $raw    = (string) curl_exec($ch);
        $status = (int) curl_getinfo($ch, CURLINFO_HTTP_CODE);
        $err    = curl_error($ch);
        curl_close($ch);

        // Parse Retry-After: accept integer seconds or HTTP-date
        $retryAfterSec = 0;
        if ($retryAfterRaw !== null) {
            if (is_numeric($retryAfterRaw)) {
                $retryAfterSec = (int) $retryAfterRaw;
            } elseif (($ts = strtotime($retryAfterRaw)) !== false) {
                $retryAfterSec = max(0, $ts - time());
            }
        }

        if ($raw === '' && $err !== '') {
            return [0, '', $err, 0];
        }

        return [$status, $raw, '', $retryAfterSec];
    }

    /** @return mixed */
    private function parse(int $status, string $raw): mixed
    {
        $data = null;
        if ($raw !== '') {
            try {
                $data = json_decode($raw, true, 512, JSON_THROW_ON_ERROR);
            } catch (\JsonException) {
                $data = $raw;
            }
        }

        if ($status >= 400) {
            $errorCode = is_array($data) ? ($data['error'] ?? null) : null;
            $message   = is_array($data) ? ($data['message'] ?? null) : null;
            // { error: "tier_insufficient", tier: "..." } 403
            // { error: "...", message: "...", code: "..." } 4xx/5xx
            throw new MelayaException(
                'Melaya API error: HTTP ' . $status . ($message ? " — {$message}" : ''),
                $status,
                is_string($errorCode) ? $errorCode : null,
                $data,
            );
        }

        // Envelope-level failure: ok === false
        if (is_array($data) && isset($data['ok']) && $data['ok'] === false) {
            $errorCode = $data['error'] ?? null;
            $message   = $data['message'] ?? null;
            throw new MelayaException(
                'Melaya API request failed' . ($message ? ": {$message}" : ''),
                $status,
                is_string($errorCode) ? $errorCode : null,
                $data,
            );
        }

        return $data;
    }
}
